osquery/tables/system/darwin/disk_encryption.mm

```diff
@@ -0,0 +1,347 @@
+      CFSTR("/System/Library/PrivateFrameworks/DiskManagement.framework"),
+      kCFURLPOSIXPathStyle,
+      true);
+  if (bundle_url == nullptr) {
+    LOG(ERROR) << "Error parsing DiskManagement bundle URL";
+    return;
+  }
+
+  auto bundle = CFBundleCreate(kCFAllocatorDefault, bundle_url);
+  CFRelease(bundle_url);
+  if (bundle == nullptr) {
+    LOG(ERROR) << "Error opening DiskManagement bundle";
+    return;
+  }
+
+  CFBundleLoadExecutable(bundle);
+
+  DASessionRef session = DASessionCreate(kCFAllocatorDefault);
+
+  if (session == nullptr) {
+    LOG(ERROR) << "Error creating DiskArbitration session";
+    CFBundleUnloadExecutable(bundle);
+    CFRelease(bundle);
+    return;
+  }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+    // DMManager *man = [DMManager sharedManager]
+    id cls = NSClassFromString(@"DMManager");
+    if (![cls respondsToSelector:@selector(sharedManager)]) {
+      LOG(ERROR) << "DMManager does not respond to sharedManager selector";
+    }
+    id man = [cls performSelector:@selector(sharedManager)];
+
+    // DMAPFS * apfs = [[DMAPFS alloc] initWithManager:man];
+    cls = NSClassFromString(@"DMAPFS");
+    if (cls == nullptr) {
+      LOG(ERROR) << "Could not load DMAPFS class";
+      CFBundleUnloadExecutable(bundle);
+      CFRelease(bundle);
+      return;
+    }
+    // Must use NSSelectorFromString or compiler complains about using alloc
+    // with ARC.
+    id apfs = [cls performSelector:NSSelectorFromString(@"alloc")];
+    if (apfs == nullptr) {
+      LOG(ERROR) << "Could not allocate DMAPFS object";
+      CFRelease(session);
+      CFBundleUnloadExecutable(bundle);
+      CFRelease(bundle);
+      return;
+    }
+    [apfs performSelector:@selector(initWithManager:) withObject:man];
+
+#pragma clang diagnostic pop
+
+    DADiskRef targetVol = DADiskCreateFromBSDName(nullptr, session, r["name"].c_str());
+
+    // err = [apfs isEncryptedVolume:targetVol encrypted:&isEncrypted];
+    char *typeEncodings;
+    asprintf(&typeEncodings,
+             "%s%s%s%s%s",
+             @encode(int), // return
+             @encode(id), // self
+             @encode(SEL), // _cmd
+             @encode(DADiskRef),
+             @encode(char*));
+    NSMethodSignature *signature = [NSMethodSignature signatureWithObjCTypes: typeEncodings];
+    free(typeEncodings);
+
+    char isEncrypted = 1;
+    char *isEncryptedPtr = &isEncrypted;
+    int err = 0;
+
+    NSInvocation *inv = [NSInvocation invocationWithMethodSignature:signature];
+    [inv setSelector:@selector(isEncryptedVolume:encrypted:)];
+    [inv setReturnValue:&err];
+    [inv setArgument:&targetVol atIndex:2];
+    [inv setArgument:&isEncryptedPtr atIndex:3];
+    [inv invokeWithTarget:apfs];
+    if (targetVol == nullptr) {
+      LOG(ERROR)
+          << "Error calling isEncryptedVolume:encrypted:, targetVol is null";
+      CFRelease((__bridge CFTypeRef)apfs);
+      CFRelease(session);
```

Comment:
Maybe this series of releases and unloading can be wrapped into a reusable lambda?